src/toil/lib/docker.py

```diff
@@ -1,280 +1,340 @@
-"""
-    Module for calling Docker. Assumes `docker` is on the PATH.
-
-    Contains two user-facing functions: dockerCall and dockerCheckOutput
-
-    Uses Toil's defer functionality to ensure containers are shutdown even in case of job or pipeline failure
-
-    Example of using dockerCall in a Toil pipeline to index a FASTA file with SAMtools:
-        def toil_job(job):
-            work_dir = job.fileStore.getLocalTempDir()
-            path = job.fileStore.readGlobalFile(ref_id, os.path.join(work_dir, 'ref.fasta')
-            parameters = ['faidx', path]
-            dockerCall(job, tool='quay.io/ucgc_cgl/samtools:latest', work_dir=work_dir, parameters=parameters)
-"""
-from builtins import str
-from builtins import map
-import base64
+from __future__ import absolute_import
 import logging
 import os
 import pipes
-import uuid
+import subprocess
+import docker
+import base64
+import time
+import requests
+
+from docker.errors import create_api_error_from_http_exception
+from docker.errors import ContainerError
+from docker.errors import ImageNotFound
+from docker.errors import APIError
+from docker.errors import NotFound
+from docker.errors import DockerException
+from docker.utils.types import LogConfig
+from docker.api.container import ContainerApiMixin
 
-from bd2k.util.exceptions import require
 from bd2k.util.retry import retry
+from docker import client
+from pwd import getpwuid
 
-from toil.lib import *
+from toil.lib import dockerPredicate
+from toil.lib import FORGO
+from toil.lib import STOP
+from toil.lib import RM
+from toil.test import timeLimit
 
 _logger = logging.getLogger(__name__)
 
 
+def dockerCheckOutput(job, *args, **kwargs):
+    """
+    Runs dockerCall().
+
+    Provided for backwards compatibility with a previous implementation that
+    used 'subprocess.check_call()' and 'subprocess.check_output()'.  This has
+    since been supplanted and all dockerCall() runs now return the same output
+    so there is no need for a separate dockerCheckOutput() function that calls
+    'subprocess.check_output()'.
+
+    See dockerCall().
+    """
+    return dockerCall(job=job, *args, **kwargs)
+
 def dockerCall(job,
                tool,
-               parameters=None,
+               parameters=[],
```

Comment:
Shouldn't have [any mutable arguments](http://docs.python-guide.org/en/latest/writing/gotchas/).

Reply:
Updated.  Thanks for the link!